File: src/wifi.c
#include "../include/wifi.h"

const char WIFI_SSID[] = "";
const char WIFI_PASSWORD[] = "";

bool done = false;
char request_body[BUF_SIZE];

int wifi_setup()
{
    cyw43_arch_init();
    cyw43_arch_enable_sta_mode();

    if (cyw43_arch_wifi_connect_timeout_ms(WIFI_SSID, WIFI_PASSWORD, CYW43_AUTH_WPA2_AES_PSK, 30000))
    {
        printf("failed to connect.\n");
        return 1;
    }

    printf("Connected.\n");
    return 0;
}

void run_tcp_server_test(void)
{
    TCP_SERVER_T *state = tcp_server_init();
    if (!state)
    {
        return;
    }
    if (!tcp_server_open(state))
    {
        DEBUG_printf("failed to open\n");
        tcp_server_result(state, -1);
        return;
    }
    while (!state->complete && !done)
    {
        cyw43_arch_poll();
        cyw43_arch_wait_for_work_until(make_timeout_time_ms(1000));
    }

    free(state);
}

static TCP_SERVER_T *tcp_server_init(void)
{
    TCP_SERVER_T *state = calloc(1, sizeof(TCP_SERVER_T));
    if (!state)
    {
        DEBUG_printf("failed to allocate state\n");
        return NULL;
    }
    return state;
}

static bool tcp_server_open(void *arg)
{
    TCP_SERVER_T *state = (TCP_SERVER_T *)arg;
    DEBUG_printf("Starting server at %s on port %u\n", ip4addr_ntoa(netif_ip4_addr(netif_list)), TCP_PORT);

    struct tcp_pcb *pcb = tcp_new_ip_type(IPADDR_TYPE_ANY);
    if (!pcb)
    {
        DEBUG_printf("failed to create pcb\n");
        return false;
    }

    err_t err = tcp_bind(pcb, NULL, TCP_PORT);
    if (err)
    {
        DEBUG_printf("failed to bind to port %u\n", TCP_PORT);
        return false;
    }

    state->server_pcb = tcp_listen_with_backlog(pcb, 1);
    if (!state->server_pcb)
    {
        DEBUG_printf("failed to listen\n");
        if (pcb)
        {
            tcp_close(pcb);
        }
        return false;
    }

    tcp_arg(state->server_pcb, state);
    tcp_accept(state->server_pcb, tcp_server_accept);

    return true;
}

static err_t tcp_server_result(void *arg, int status)
{
    TCP_SERVER_T *state = (TCP_SERVER_T *)arg;
    if (status == 0)
    {
        DEBUG_printf("test success\n");
    }
    else
    {
        DEBUG_printf("test failed %d\n", status);
    }
    state->complete = true;
    return tcp_server_close(arg);
}

static err_t tcp_server_accept(void *arg, struct tcp_pcb *client_pcb, err_t err)
{
    TCP_SERVER_T *state = (TCP_SERVER_T *)arg;
    if (err != ERR_OK || client_pcb == NULL)
    {
        DEBUG_printf("Failure in accept\n");
        tcp_server_result(arg, err);
        return ERR_VAL;
    }
    DEBUG_printf("Client connected\n");

    state->client_pcb = client_pcb;
    tcp_arg(client_pcb, state);
    tcp_sent(client_pcb, tcp_server_sent);
    tcp_recv(client_pcb, tcp_server_recv);
    tcp_poll(client_pcb, tcp_server_poll, POLL_TIME_S * 2);
    tcp_err(client_pcb, tcp_server_err);

    return ERR_OK;
}

static err_t tcp_server_close(void *arg)
{
    TCP_SERVER_T *state = (TCP_SERVER_T *)arg;
    err_t err = ERR_OK;
    if (state->client_pcb != NULL)
    {
        tcp_arg(state->client_pcb, NULL);
        tcp_poll(state->client_pcb, NULL, 0);
        tcp_sent(state->client_pcb, NULL);
        tcp_recv(state->client_pcb, NULL);
        tcp_err(state->client_pcb, NULL);
        err = tcp_close(state->client_pcb);
        if (err != ERR_OK)
        {
            DEBUG_printf("close failed %d, calling abort\n", err);
            tcp_abort(state->client_pcb);
            err = ERR_ABRT;
        }
        state->client_pcb = NULL;
    }
    if (state->server_pcb)
    {
        tcp_arg(state->server_pcb, NULL);
        tcp_close(state->server_pcb);
        state->server_pcb = NULL;
    }
    return err;
}

static err_t tcp_server_sent(void *arg, struct tcp_pcb *tpcb, u16_t len)
{
    TCP_SERVER_T *state = (TCP_SERVER_T *)arg;
    DEBUG_printf("tcp_server_sent %u\n", len);
    state->sent_len += len;

    if (state->sent_len >= BUF_SIZE)
    {
        state->recv_len = 0;
        DEBUG_printf("Waiting for buffer from client\n");
    }

    return ERR_OK;
}

err_t tcp_server_send_data(void *arg, struct tcp_pcb *tpcb)
{
    TCP_SERVER_T *state = (TCP_SERVER_T *)arg;

    char response[BUF_SIZE];

    done ? handle_post_request(response) : handle_options_request(response);

    strcpy(state->buffer_sent, response);

    DEBUG_printf(CYAN "Pico responded with: %.*s\n" RESET, BUF_SIZE, state->buffer_sent);
    DEBUG_printf(MAGENTA "********************************************************\n" RESET);

    state->sent_len = 0;
    cyw43_arch_lwip_check();
    err_t err = tcp_write(tpcb, state->buffer_sent, BUF_SIZE, TCP_WRITE_FLAG_COPY);
    if (err != ERR_OK)
    {
        DEBUG_printf("Failed to write data %d\n", err);
        return tcp_server_result(arg, -1);
    }
    return ERR_OK;
}

err_t tcp_server_recv(void *arg, struct tcp_pcb *tpcb, struct pbuf *p, err_t err)
{
    TCP_SERVER_T *state = (TCP_SERVER_T *)arg;
    if (!p)
    {
        return tcp_server_result(arg, -1);
    }

    cyw43_arch_lwip_check();
    if (p->tot_len > 0)
    {
        DEBUG_printf("tcp_server_recv %d/%d err %d\n", p->tot_len, state->recv_len, err);

        // Receive the buffer
        const uint16_t buffer_left = BUF_SIZE;
        state->recv_len = pbuf_copy_partial(p, state->buffer_recv,
                                            p->tot_len > buffer_left ? buffer_left : p->tot_len, 0);
        tcp_recved(tpcb, p->tot_len);

        done = strncmp(state->buffer_recv, "POST", 4) == 0;

        if (done)
        {
            memcpy(request_body, state->buffer_recv, BUF_SIZE);
            request_body[BUF_SIZE - 1] = '\0';
        }

        DEBUG_printf(YELLOW "Received request: %s\n" RESET, state->buffer_recv);

        return tcp_server_send_data(arg, tpcb);
    }
    pbuf_free(p);

    return ERR_OK;
}

static err_t tcp_server_poll(void *arg, struct tcp_pcb *tpcb)
{
    DEBUG_printf("tcp_server_poll_fn\n");
    return tcp_server_result(arg, -1); // no response is an error?
}

static void tcp_server_err(void *arg, err_t err)
{
    if (err != ERR_ABRT)
    {
        DEBUG_printf("tcp_client_err_fn %d\n", err);
        tcp_server_result(arg, err);
    }
}

void handle_post_request(char *response)
{
    const char *response_body = "{ \"message\": \"Pico W received location and compass readings successfully 🎉🎉\" }";
    int content_length = strlen(response_body);

    snprintf(response, BUF_SIZE,
             "HTTP/1.1 200 OK\r\n"
             "Access-Control-Allow-Origin: *\r\n"
             "Access-Control-Allow-Methods: GET, POST, OPTIONS\r\n"
             "Access-Control-Allow-Headers: Content-Type, Accept\r\n"
             "Access-Control-Max-Age: 3600\r\n"
             "Content-Type: application/json\r\n"
             "Content-Length: %d\r\n\r\n%s",
             content_length, response_body);
}

void handle_options_request(char *response)
{
    snprintf(response, BUF_SIZE,
             "HTTP/1.1 204 No Content\r\n"
             "Access-Control-Allow-Origin: *\r\n"
             "Access-Control-Allow-Methods: GET, POST, OPTIONS\r\n"
             "Access-Control-Allow-Headers: Content-Type, Accept\r\n"
             "Access-Control-Max-Age: 3600\r\n\r\n");
}

File: src/wifi.py
import network
import socket
import asyncio
from time import sleep
import json

class PicoWWebReceiver:
    """
    Initialize the Pico W web receiver
    
    :param ssid: WiFi network name
    :param password: WiFi network password
    """
    def __init__(self, ssid, password):
        self.wlan = network.WLAN(network.STA_IF)
        self.ssid = ssid
        self.password = password
        
        self.init_wifi()
        
    def init_wifi(self):
        self.wlan.active(True)
        self.wlan.connect(self.ssid, self.password)
        
        max_wait = 10
        while max_wait > 0:
            if self.wlan.status() < 0 or self.wlan.status() >= 3:
                break
            max_wait -= 1
            print('waiting for connection...')
            sleep(1)
        
        if self.wlan.status() != 3:
            raise RuntimeError('network connection failed')
        else:
            status = self.wlan.ifconfig()
            print('ip = ' + status[0])
        
    async def handle_client(self, reader, writer):
        try:
            request_line = await reader.readline()
            request_method = str(request_line, 'utf-8').split()[0] if request_line else ''
                
            print(f"Received {request_method} request")

            cors_headers = (
                "HTTP/1.1 200 OK\r\n"
                "Access-Control-Allow-Origin: *\r\n"
                "Access-Control-Allow-Methods: GET, POST, OPTIONS\r\n"
                "Access-Control-Allow-Headers: Content-Type, Accept\r\n"
                "Access-Control-Max-Age: 3600\r\n"
            )

            # Handle OPTIONS preflight
            if request_method == 'OPTIONS':
                response = (
                    f"{cors_headers}"
                    "Content-Length: 0\r\n"
                    "\r\n"
                )
                writer.write(response.encode())
                await writer.drain()
                return

            if request_method == 'POST':
                content_length = 0
                while True:
                    header = await reader.readline()
                    if header == b'\r\n':
                        break
                    if header.lower().startswith(b'content-length:'):
                        content_length = int(header.decode().split(':')[1].strip())

                body = await reader.read(content_length)                
                parsed_data = json.loads(body.decode('utf-8'))
                print('Received data:', parsed_data)

                # Prepare response
                response_body = json.dumps({
                    'message': 'Pico W received location and compass readings successfully 🎉🎉'
                })

                response = (
                    f"{cors_headers}"
                    f"Content-Type: application/json\r\n"
                    f"Content-Length: {len(response_body)}\r\n"
                    "\r\n"
                    f"{response_body}"
                )

                writer.write(response.encode())
                await writer.drain()
        
        except Exception as e:
            print(f"Error handling client: {e}")
            # Send error response
            error_response = (
                f"{cors_headers}"
                "Content-Type: application/json\r\n"
                f"Content-Length: {len(str(e))}\r\n"
                "\r\n"
                f"{{'error': '{e}'}}"
            )
            writer.write(error_response.encode())

        finally:
            writer.close()
            await writer.wait_closed()


async def main():
    WIFI_SSID = ''
    WIFI_PASSWORD = ''
    
    receiver = PicoWWebReceiver(WIFI_SSID, WIFI_PASSWORD)

    server = asyncio.start_server(receiver.handle_client, "0.0.0.0", 80)
    asyncio.create_task(server)

loop = asyncio.get_event_loop()
loop.create_task(main())

try:
    loop.run_forever()
except Exception as e:
    print('Error occured: ', e)
except KeyboardInterrupt:
    print('Program Interrupted by the user')
